File: collect_app/src/main/java/org/odk/collect/android/preferences/PrefUtils.java
package org.odk.collect.android.preferences;

import android.content.Context;

import androidx.preference.ListPreference;

import org.odk.collect.android.application.Collect;
import org.odk.collect.android.injection.DaggerUtils;
import org.odk.collect.shared.settings.Settings;

import java.util.Arrays;

public final class PrefUtils {
    private PrefUtils() { }  // prevent instantiation of this utility class

    public static Settings getSharedPrefs() {
        return DaggerUtils.getComponent(Collect.getInstance()).settingsProvider().getUnprotectedSettings();
    }

    public static ListPreference createListPref(
        Context context, String key, String title, int[] labelIds, String[] values) {
        String[] labels = new String[labelIds.length];
        for (int i = 0; i < labels.length; i++) {
            labels[i] = context.getString(labelIds[i]);
        }
        return createListPref(context, key, title, labels, values);
    }

    private static ListPreference createListPref(
        Context context, String key, String title, String[] labels, String[] values) {
        ensurePrefHasValidValue(key, values);
        ListPreference pref = new ListPreference(context);
        pref.setKey(key);
        pref.setPersistent(true);
        pref.setTitle(title);
        pref.setDialogTitle(title);
        pref.setEntries(labels);
        pref.setEntryValues(values);
        pref.setSummary("%s");
        return pref;
    }

    private static void ensurePrefHasValidValue(String key, String[] validValues) {
        Settings prefs = getSharedPrefs();
        String value = prefs.getString(key);
        if (Arrays.asList(validValues).indexOf(value) < 0) {
            if (validValues.length > 0) {
                prefs.save(key, validValues[0]);
            } else {
                prefs.remove(key);
            }
        }
    }

    /**
     * Gets an integer value from the shared preferences.  If the preference has
     * a string value, attempts to convert it to an integer.  If the preference
     * is not found or is not a valid integer, returns the defaultValue.
     */
    public static int getInt(String key, int defaultValue) {
        Object value = getSharedPrefs().getAll().get(key);
        if (value instanceof Integer) {
            return (Integer) value;
        }
        if (value instanceof String) {
            try {
                return Integer.parseInt((String) value);
            } catch (NumberFormatException e) { /* ignore */ }
        }
        return defaultValue;
    }
}

File: androidshared/src/main/java/org/odk/collect/androidshared/ui/PrefUtils.kt
package org.odk.collect.androidshared.ui

import android.content.Context
import androidx.preference.ListPreference
import org.odk.collect.shared.settings.Settings

object PrefUtils {

    @JvmStatic
    fun createListPref(
        context: Context,
        key: String,
        title: String,
        labelIds: IntArray,
        values: Array<String>,
        settings: Settings,
    ): ListPreference {
        val labels: Array<String?> = labelIds.map { context.getString(it) }.toTypedArray()
        return createListPref(context, key, title, labels, values, settings)
    }

    /**
     * Gets an integer value from the shared preferences. If the preference has
     * a string value, attempts to convert it to an integer. If the preference
     * is not found or is not a valid integer, returns the defaultValue.
     */
    @JvmStatic
    fun getInt(key: String?, defaultValue: Int, settings: Settings): Int {
        val value: Any? = settings.getAll()[key]
        if (value is Int) {
            return value
        }

        if (value is String) {
            try {
                return Integer.parseInt(value)
            } catch (e: NumberFormatException) {
                // ignore
            }
        }

        return defaultValue
    }

    private fun createListPref(
        context: Context,
        key: String,
        title: String,
        labels: Array<String?>,
        values: Array<String>,
        settings: Settings,
    ): ListPreference {
        ensurePrefHasValidValue(key, values, settings)
        return ListPreference(context).also {
            it.key = key
            it.isPersistent = true
            it.title = title
            it.dialogTitle = title
            it.entries = labels
            it.entryValues = values
            it.summary = "%s"
        }
    }

    private fun ensurePrefHasValidValue(
        key: String,
        validValues: Array<String>,
        settings: Settings,
    ) {
        val value = settings.getString(key)
        if (validValues.indexOf(value) < 0) {
            if (validValues.isNotEmpty()) {
                settings.save(key, validValues[0])
            } else {
                settings.remove(key)
            }
        }
    }
}
